File: dataStructures/linkedList/linkedlist.c
#include "linkedlist.h"

Node *createNode(int data)
{
    Node *newNode = malloc(sizeof(Node));

    if (newNode == NULL)
        return NULL;

    newNode->data = data;
    newNode->next = NULL;

    return newNode;
}

/*  */
List *createList()
{

    List *newList = malloc(sizeof(List));

    if (newList == NULL)
        return NULL;

    newList->head = NULL;

    return newList;
}

void insertAtLast(List *l, int data)
{
    if (l->head == NULL)
        l->head = createNode(data);
    else
    {
        Node *temp = l->head;

        while (temp->next != NULL)
            temp = temp->next;

        temp->next = createNode(data);
    }
}

void insertAtFirst(List *l, int data)
{
    if (l->head == NULL)
        l->head = createNode(data);
    else
    {
        Node *temp = createNode(data);
        temp->next = l->head;
        l->head = temp;
    }
}

void insertSorted(List* l, int data){

    Node* newNode = createNode(data);
    
    if(l->head == NULL){
        l->head = newNode;
    }
    else{
        Node* temp = l->head;

        if(l->head->data > data){
            newNode->next = l->head;
            l->head = newNode;
            return;
        }

        while(temp->next != NULL && temp->next->data < data ){
            temp = temp->next;
        }
        
        if(temp->next == NULL){
            temp->next = newNode;
        }
        else{
            Node* temp2 = temp->next;
            temp->next = newNode;
            newNode->next = temp2;
        }
    }
}


void removeLast(List *l)
{
    if(l->head != NULL){
        
        Node* temp = l->head;
        
        while(temp->next->next != NULL)
            temp = temp->next;
        
        free(temp->next);
        temp->next = NULL;


    }
}

void removeFirst(List *l)
{
    if (l->head != NULL)
    {
        Node *temp = l->head;
        l->head = l->head->next;
        free(temp);
    }
}

void printList(List *l)
{
    Node *temp = l->head;

    if (l->head == NULL)
        return;

    printf("[ ");
    while (temp->next != NULL)
    {
        printf("%d, ", temp->data);
        temp = temp->next;
    }

    printf("%d ]\n", temp->data);
}

File: javascript/linked-list/linked-list.js
function Node(value) {
  this.value = value;
  this.next = null;
}

function LinkedList() {
  this.head = null;
}

// O(n)
function printList(linkedList) {
  let temp = linkedList.head;

  while (temp !== null) {
    process.stdout.write(`${temp.value}, `);
    temp = temp.next;
  }

  process.stdout.write("\n");
}

// O(1)
function insertAtFirst(linkedList, value) {
  let node = new Node(value);
  node.next = linkedList.head;
  linkedList.head = node;
}

// O(n)
function insertAtLast(linkedList, value) {
  let node = new Node(value);

  if (linkedList.head === null) {
    linkedList.head = node;
    return;
  }

  let temp = linkedList.head;

  while (temp.next !== null) {
    temp = temp.next;
  }

  temp.next = node;
}

// O(n)
function searchValue(linkedList, value) {
  let temp = linkedList.head;

  if (temp === null) {
    return false;
  }

  while (temp !== null) {
    if (temp.value === value) {
      return true;
    }

    temp = temp.next;
  }
  return false;
}

(() => {
  const linkedList = new LinkedList();

  insertAtLast(linkedList, 50);
  insertAtLast(linkedList, 10);
  insertAtLast(linkedList, 30);
  insertAtFirst(linkedList, 1);

  printList(linkedList);

  console.log(searchValue(linkedList, 1));
  console.log(searchValue(linkedList, 100));
  console.log(searchValue(linkedList, 30));
})();
